openshift-kube-apiserver/admission/customresourcevalidation/authentication/validate_authentication.go

```diff
@@ -121,14 +166,264 @@ func validateAuthenticationSpec(spec configv1.AuthenticationSpec) field.ErrorLis
 				spec.WebhookTokenAuthenticator, fmt.Sprintf("this field cannot be set with the %q .spec.type", spec.Type),
 			))
 		}
-
 	}
 
 	errs = append(errs, crvalidation.ValidateConfigMapReference(specField.Child("oauthMetadata"), spec.OAuthMetadata, false)...)
 
-	return errs
+	// Perform External OIDC Provider related validations
+	// ----------------
+
+	// There is currently no guarantee that these fields are not set when the spec.Type is != OIDC.
+	// To ensure we are enforcing approriate admission validations at all times, just always iterate through the list
+	// of OIDC Providers and perform the validations.
+	// If/when the openshift/api admission validations are updated to enforce that this field is not configured
+	// when Type != OIDC, this loop should be a no-op due to an empty list.
+	for i, provider := range spec.OIDCProviders {
+		results := validateOIDCProvider(ctx, specField.Child("oidcProviders").Index(i), cel, provider)
+		var expressionCost uint64 = 0
+		excessiveCosts := []string{}
+		for _, result := range results {
+			if result.internalErr != nil {
```

Comment:
Is there some way to structure this so creating the field errors is very close to the place where the error is generated? If the slow compilation logging and warning is handled next to compilation, there's still a bunch of plumbing involved solely to support a cost warning.

Maybe flip things around and pass a costRecorder down from `validateOIDCProvider`, call `RecordCost(*field.Path, uint64)` on it at the bottom, then once you've returned to `validateOIDCProvider`, look at the accumulated state in costRecorder and call AddWarning(...) if necessary? Then all the validateFoo functions can return the conventional `field.ErrorList`.